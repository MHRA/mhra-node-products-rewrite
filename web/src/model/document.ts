export interface IDocument {
  activeSubstances: string[];
  context: string;
  created: string;
  docType: string;
  fileSize: string;
  name: string;
  product: string;
  url: string;
}

export interface IDocuments {
  count: number;
  documents: IDocument[];
}

export interface IBmgfReport {
  activeSubstances: string[];
  context: string;
  fileName: string;
  fileSize: string;
  fileUrl: string;
  title: string;
  products: string[];
  matrices: string[];
  pbpkModels: string[];
  summary: string;
  url: string;
  pregnancyTrimesters: string[];
  plNumbers: string[];
}

export interface IBmgfReports {
  count: number;
  reports: IBmgfReport[];
}
