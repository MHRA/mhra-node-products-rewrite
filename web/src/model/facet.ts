export interface IFacet {
  name: string;
  count?: number;
}
