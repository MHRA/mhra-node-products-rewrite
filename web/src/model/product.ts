export interface IProduct {
  name: string;
  count?: number;
}
