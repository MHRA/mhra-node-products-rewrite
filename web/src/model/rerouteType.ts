export enum RerouteType {
  CheckboxSelected,
  Other,
}
