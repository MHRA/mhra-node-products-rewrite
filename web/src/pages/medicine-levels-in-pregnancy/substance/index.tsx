import { NextPage } from 'next';
import { useRouter } from 'next/router';
import React, { useEffect } from 'react';

import { BmgfPage } from '../../../components/page';
import SearchResults from '../../../components/bmgf/search-results';
import SearchWrapper from '../../../components/bmgf/search-wrapper';
import { DrugStructuredData } from '../../../components/structured-data';
import { useLocalStorage } from '../../../hooks';
import { IBmgfReport } from '../../../model/document';
import { getLoader } from '../../../services/loaders/medicine-levels-in-pregnancy/substance-loader';
import Events from '../../../services/events';
import { parsePage } from '../../../services/querystring-interpreter';

const pageSize = 10;
const substancePath = '/medicine-levels-in-pregnancy/substance';

const App: NextPage = () => {
  const [storageAllowed, setStorageAllowed] = useLocalStorage(
    'allowStorage',
    false,
  );
  const [reports, setReports] = React.useState<IBmgfReport[]>([]);
  const [substanceName, setSubstanceName] = React.useState('');
  const [count, setCount] = React.useState(0);
  const [pageNumber, setPageNumber] = React.useState(1);
  const [isLoading, setIsLoading] = React.useState(true);
  const [errorFetchingResults, setErrorFetchingResults] = React.useState(false);

  const useGraphQl: boolean = process.env.USE_GRAPHQL === 'true';

  const router = useRouter();
  const {
    query: { substance: substanceQS, page: pageQS },
  } = router;

  useEffect(() => {
    if (!substanceQS) {
      return;
    }
    const substance = substanceQS.toString();
    const page = pageQS ? parsePage(pageQS) : 1;
    setSubstanceName(substance);
    setPageNumber(page);
    setReports([]);
    setCount(0);
    setIsLoading(true);
    setErrorFetchingResults(false);

    getLoader(useGraphQl)
      .load({
        name: substance,
        page,
        pageSize,
      })
      .then(({ reports, count }) => {
        setReports(reports);
        setCount(count);
        setIsLoading(false);
      })
      .catch((e) => setErrorFetchingResults(true));

    Events.viewPbpkResultsForSubstance({
      substance,
      pageNo: page,
    });
  }, [substanceQS, pageQS]);

  useEffect(() => {
    if (window) {
      window.scrollTo(0, 0);
    }
  }, []);

  const reroutePage = (substanceName: string, page: number) => {
    const query = {
      substance: substanceName,
      page,
    };
    router.push({
      pathname: substancePath,
      query,
    });
  };

  const handlePageChange = async (page: number) => {
    reroutePage(substanceName, page);
  };

  return (
    <BmgfPage
      title="Medicine levels in pregnancy"
      metaTitle="Medicine levels in pregnancy | Substance results"
      storageAllowed={storageAllowed}
      setStorageAllowed={setStorageAllowed}
    >
      <SearchWrapper initialSearchValue="">
        <SearchResults
          reports={reports}
          showingResultsForTerm={substanceName}
          resultCount={count}
          page={pageNumber}
          pageSize={pageSize}
          searchTerm={substanceName}
          handlePageChange={handlePageChange}
          isLoading={isLoading}
          errorFetchingResults={errorFetchingResults}
        />
        <DrugStructuredData drugName={substanceName} />
      </SearchWrapper>
    </BmgfPage>
  );
};

export default App;
