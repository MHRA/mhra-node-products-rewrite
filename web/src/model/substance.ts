import { IProduct } from './product';

export interface ISubstance extends IProduct {
  products?: IProduct[];
}

export function isSubstance(obj: any): obj is ISubstance {
  return obj.products !== undefined && obj.count !== undefined;
}

export function isIndex(obj: any): obj is ISubstance {
  return obj.products === undefined && obj.count === undefined;
}
