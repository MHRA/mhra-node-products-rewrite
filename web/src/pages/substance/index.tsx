import { NextPage } from 'next';
import { useRouter } from 'next/router';
import React, { useEffect } from 'react';
import Page from '../../components/page';
import ProductList from '../../components/product-list/index';
import SearchWrapper from '../../components/search-wrapper';
import {
  DrugListStructuredData,
  SubstanceStructuredData,
} from '../../components/structured-data';
import { useLocalStorage } from '../../hooks';
import { IProduct } from '../../model/product';
import Events from '../../services/events';
import { getLoader } from '../../services/loaders/products/products-index-loader';

const App: NextPage = () => {
  const [storageAllowed, setStorageAllowed] = useLocalStorage(
    'allowStorage',
    false,
  );
  const [products, setProducts] = React.useState<IProduct[]>([]);
  const [substanceName, setSubstanceName] = React.useState('');
  const [isLoading, setIsLoading] = React.useState(true);
  const [errorFetchingResults, setErrorFetchingResults] = React.useState(false);
  const useGraphQl: boolean = process.env.USE_GRAPHQL === 'true';
  const router = useRouter();
  const {
    query: { substance: queryQS },
  } = router;

  useEffect(() => {
    if (!queryQS) {
      return;
    }
    const substanceName = queryQS.toString();
    setSubstanceName(substanceName);

    setProducts([]);
    setIsLoading(true);
    setErrorFetchingResults(false);

    const loader = getLoader(useGraphQl);

    loader
      .load(substanceName)
      .then((products) => {
        setProducts(products);
        setIsLoading(false);
      })
      .catch((e) => setErrorFetchingResults(true));

    Events.viewProductsForSubstance(substanceName);
  }, [queryQS]);

  useEffect(() => {
    if (window) {
      window.scrollTo(0, 0);
    }
  }, []);

  return (
    <Page
      title="Products"
      metaTitle="Products | Substance"
      storageAllowed={storageAllowed}
      setStorageAllowed={setStorageAllowed}
    >
      <SearchWrapper initialSearchValue="">
        <ProductList
          title={substanceName}
          products={products}
          errorFetchingResults={errorFetchingResults}
          isLoading={isLoading}
        />
        <SubstanceStructuredData substanceName={substanceName} />
        {products && products.length ? (
          <DrugListStructuredData
            drugNames={products.map((product) => product.name)}
          />
        ) : (
          <></>
        )}
      </SearchWrapper>
    </Page>
  );
};

export default App;
