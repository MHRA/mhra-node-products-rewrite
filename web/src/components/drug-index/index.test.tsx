import renderer from 'react-test-renderer';
import React from 'react';
import { ISubstance } from '../../model/substance';
import { IProduct } from '../../model/product';
import DrugIndex, { IndexType } from './index';

describe(DrugIndex, () => {
  it('should render horizontal items', () => {
    const component = renderer
      .create(
        <DrugIndex
          title={'Coffee'}
          indexType={IndexType.Horizontal}
          items={[
            { name: 'A' },
            { name: 'B' },
            { name: 'C' },
            { name: 'D' },
            { name: 'E' },
            { name: 'F' },
            { name: 'G' },
            { name: 'H' },
            { name: 'I' },
            { name: 'J' },
            { name: 'K' },
            { name: 'L' },
            { name: 'M' },
            { name: 'N' },
            { name: 'O' },
            { name: 'P' },
            { name: 'Q' },
            { name: 'R' },
            { name: 'S' },
            { name: 'T' },
            { name: 'U' },
            { name: 'V' },
            { name: 'W' },
            { name: 'X' },
            { name: 'Y' },
            { name: 'Z' },
            { name: '0' },
            { name: '1' },
            { name: '2' },
            { name: '3' },
            { name: '4' },
            { name: '5' },
            { name: '6' },
            { name: '7' },
            { name: '8' },
            { name: '9' },
          ]}
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it('should render substances', () => {
    const product: IProduct = { name: 'Ibuprofen gel', count: 1 };
    const substance: ISubstance = {
      name: 'Ibuprofen',
      products: [product],
      count: 1,
    };
    const substances = [substance];
    const component = renderer
      .create(
        <DrugIndex
          title={'Coffee'}
          items={substances}
          indexType={IndexType.SubstancesIndex}
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it('should render error message', () => {
    const component = renderer
      .create(
        <DrugIndex
          title={'Coffee'}
          items={[]}
          indexType={IndexType.ProductsIndex}
          errorFetchingResults
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it('should render loading message', () => {
    const component = renderer
      .create(
        <DrugIndex
          title={'Coffee'}
          items={[]}
          indexType={IndexType.ProductsIndex}
          isLoading
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
