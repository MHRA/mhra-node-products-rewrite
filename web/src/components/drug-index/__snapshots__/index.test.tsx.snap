// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugIndex should render error message 1`] = `
<nav
  className="sc-bczRLJ hrwWmM"
>
  <p
    className="sc-gsnTZi dPidEJ"
  >
    Sorry - the site is experiencing technical issues right now. Please try again later.
  </p>
</nav>
`;

exports[`DrugIndex should render horizontal items 1`] = `
<nav
  className="sc-bczRLJ hrwWmM"
>
  <p
    className="horizontal"
  >
    Coffee
  </p>
  <ul
    className="horizontal"
  >
    <li
      className=""
    >
      <a
        href="/substance-index?letter=A"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        A
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=B"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        B
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=C"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        C
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=D"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        D
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=E"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        E
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=F"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        F
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=G"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        G
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=H"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        H
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=I"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        I
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=J"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        J
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=K"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        K
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=L"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        L
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=M"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        M
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=N"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        N
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=O"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        O
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=P"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        P
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=Q"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Q
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=R"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        R
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=S"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        S
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=T"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        T
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=U"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        U
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=V"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        V
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=W"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        W
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=X"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        X
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=Y"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Y
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=Z"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Z
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=0"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        0
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=1"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        1
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        2
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=3"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        3
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=4"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        4
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=5"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        5
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=6"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        6
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=7"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        7
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=8"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        8
         
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/substance-index?letter=9"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        9
         
      </a>
    </li>
  </ul>
</nav>
`;

exports[`DrugIndex should render loading message 1`] = `
<nav
  className="sc-bczRLJ hrwWmM"
>
  <h2>
    Coffee
  </h2>
  <ul
    className=""
  >
    <li>
      Loading results...
    </li>
  </ul>
</nav>
`;

exports[`DrugIndex should render substances 1`] = `
<nav
  className="sc-bczRLJ hrwWmM"
>
  <h2>
    Coffee
  </h2>
  <ul
    className=""
  >
    <li
      className="substance-name"
    >
      <a
        href="/substance?substance=Ibuprofen"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Ibuprofen
         
        (
        1
         
        file
        )
      </a>
    </li>
  </ul>
</nav>
`;
